# Hacker News Guided Project - DataQuest

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.




We'll compare these two types of posts to determine the following:

- Do Ask HN or Show HN receive more comments on average?

- Do posts created at a certain time receive more comments on average?

Let's start by importing the libraries we need and reading the data set into a list of lists.


In [10]:
from csv import reader

In [11]:
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
h = list(read_file)
headers = h[0]
hn = h[1:]

In [12]:
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [15]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn') == True:
        ask_posts.append(row)
    elif title.lower().startswith('show hn') == True:
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
    

1744
1162
17194


In [16]:
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments += int(row[4])
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

total_show_comments = 0
for row in show_posts:
    total_show_comments += int(row[4])
avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)

14.038417431192661
10.31669535283993


The average comments for ask posts 14.0 while that for show posts are 10.3. Therefore ask posts recieve more comments than show posts.

In [17]:
import datetime as dt

In [22]:
result_list = []
for row in ask_posts:
    result_list.append([row[6], row[4]])
    
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = dt.datetime.strptime(row[0], '%m/%d/%Y %H:%M')
    date = date.strftime('%H')
    if date not in counts_by_hour:
        counts_by_hour[date] = 1
        comments_by_hour[date] = int(row[1])
    else:
        counts_by_hour[date] += 1
        comments_by_hour[date] += int(row[1])

In [24]:
print(counts_by_hour)
print(comments_by_hour)

{'23': 68, '05': 46, '07': 34, '20': 80, '11': 58, '09': 45, '01': 60, '18': 109, '19': 110, '12': 73, '04': 47, '15': 116, '16': 108, '06': 44, '00': 55, '21': 109, '10': 59, '08': 48, '03': 54, '17': 100, '13': 85, '22': 71, '02': 58, '14': 107}
{'23': 543, '05': 464, '07': 267, '20': 1722, '11': 641, '09': 251, '01': 683, '18': 1439, '19': 1188, '12': 687, '04': 337, '15': 4477, '16': 1814, '06': 397, '00': 447, '21': 1745, '10': 793, '08': 492, '03': 421, '17': 1146, '13': 1253, '22': 479, '02': 1381, '14': 1416}


In [35]:
avg_by_hour = []
for comment in comments_by_hour:
    avg_by_hour.append([comment, comments_by_hour[comment]/len(comments_by_hour)])

In [36]:
print(avg_by_hour)

[['23', 22.625], ['05', 19.333333333333332], ['07', 11.125], ['20', 71.75], ['11', 26.708333333333332], ['09', 10.458333333333334], ['01', 28.458333333333332], ['18', 59.958333333333336], ['19', 49.5], ['12', 28.625], ['04', 14.041666666666666], ['15', 186.54166666666666], ['16', 75.58333333333333], ['06', 16.541666666666668], ['00', 18.625], ['21', 72.70833333333333], ['10', 33.041666666666664], ['08', 20.5], ['03', 17.541666666666668], ['17', 47.75], ['13', 52.208333333333336], ['22', 19.958333333333332], ['02', 57.541666666666664], ['14', 59.0]]


In [38]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)

[[22.625, '23'], [19.333333333333332, '05'], [11.125, '07'], [71.75, '20'], [26.708333333333332, '11'], [10.458333333333334, '09'], [28.458333333333332, '01'], [59.958333333333336, '18'], [49.5, '19'], [28.625, '12'], [14.041666666666666, '04'], [186.54166666666666, '15'], [75.58333333333333, '16'], [16.541666666666668, '06'], [18.625, '00'], [72.70833333333333, '21'], [33.041666666666664, '10'], [20.5, '08'], [17.541666666666668, '03'], [47.75, '17'], [52.208333333333336, '13'], [19.958333333333332, '22'], [57.541666666666664, '02'], [59.0, '14']]


In [42]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap[:5])

[[186.54166666666666, '15'], [75.58333333333333, '16'], [72.70833333333333, '21'], [71.75, '20'], [59.958333333333336, '18']]


In [43]:
for avg in sorted_swap[:5]:
    hour = dt.datetime.strptime(avg[1], '%H')
    hour = hour.strftime('%H:%M')
    temp = '{}: {:.2f} average comments per post'
    average = temp.format(hour, avg[0])
    print(average)

15: 186.54 average comments per post
16: 75.58 average comments per post
21: 72.71 average comments per post
20: 71.75 average comments per post
18: 59.96 average comments per post
